File: backend/app/agent_builder/schemas/tracking.py
"""HITL-07 申请人追踪页 Pydantic v2 响应模型。

CONTEXT §申请人追踪页隐私：
- 申请人能看到审批人姓名 + 决策 + 时间（不能看 IP / UA）
- admin 能看全（含 NET-05 审计字段 IP/UA）

设计参考：Dify api/controllers/console/app/workflow_run.py + WorkflowPauseDetailsResponse
（详见 docs/reading-dify-03-08-tracking-page-2026-05-17.md §7）
"""
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


# ── 单条决策记录（脱敏前后字段一致；脱敏时 ip/ua 置 None）────────────────────


class TrackingRecord(BaseModel):
    """一条审批决策记录。

    申请人视角：actor_email + actor_name + action + reason + ts（ip/ua=None）
    admin 视角：完整字段（含 ip/ua）
    """

    actor_email: str = Field("", description="审批人邮箱")
    actor_name: str | None = Field(None, description="审批人显示名")
    action: str = Field(..., description="决策动作：submit/approve/return/reject/escalate")
    reason: str = Field("", description="决策理由（可选）")
    ts: str = Field(..., description="决策时间（ISO8601 字符串）")
    ip: str | None = Field(
        None,
        description="决策来源 IP（仅 admin 可见；申请人视角为 None）",
    )
    ua: str | None = Field(
        None,
        description="决策来源 User-Agent（仅 admin 可见；申请人视角为 None）",
    )


# ── 申请人信息（简化）─────────────────────────────────────────────────────────


class TrackingApplicant(BaseModel):
    """实例申请人摘要。"""

    id: UUID
    name: str | None = Field(None, description="申请人显示名（缺失时回退 email 前缀）")


# ── 当前节点（节点可视化要求 — user feedback_node_visualization）────────────


class TrackingCurrentActor(BaseModel):
    """当前节点等待的 actor。"""

    email: str | None = Field(None, description="当前 actor 邮箱")
    name: str | None = Field(None, description="当前 actor 显示名")


class TrackingCurrentNode(BaseModel):
    """当前正在等待的节点（active 节点高亮 + 倒计时）。

    None → 实例已终态（completed/failed/aborted），无 current_node。
    """

    id: str = Field(..., description="DSL 中的节点 ID（如 hitl_review_1）")
    title: str | None = Field(None, description="节点 DSL 配置中的 title")
    status: str = Field(..., description="状态：pending/waiting_human/in_review")
    node_type: str = Field(..., description="节点类型：hitl/llm/tool/...")
    actor: TrackingCurrentActor | None = Field(None, description="当前等待的 actor")
    deadline_at: str | None = Field(None, description="截止时间（ISO8601 字符串）")
    started_at: str | None = Field(None, description="节点开始时间")


# ── 顶层响应 ──────────────────────────────────────────────────────────────────


class TrackingResponse(BaseModel):
    """申请人追踪页响应。

    CONTEXT §申请人追踪页隐私：
    - records 已在 service 层按用户角色脱敏（admin 见全；申请人 ip/ua=None）
    - current_node None 表示实例已终态
    """

    instance_id: UUID
    status: str = Field(..., description="实例状态：pending/running/completed/failed/aborted")
    workflow_id: UUID
    applicant: TrackingApplicant
    current_node: TrackingCurrentNode | None = Field(
        None,
        description="当前正在等待的节点；终态实例为 None",
    )
    records: list[TrackingRecord] = Field(
        default_factory=list,
        description="所有节点聚合后的决策时间线，按 ts ASC 排序",
    )
    created_at: datetime
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}

File: web/src/lib/types/tracking.ts
/**
 * HITL-07 申请人追踪页类型定义
 *
 * 与后端 schemas/tracking.py Pydantic 模型对齐：
 *  - records ip/ua: 申请人视角为 null（service 层脱敏）
 *  - current_node: 终态实例时为 null
 */

import type { InstanceStatus } from '@/lib/types/instance';

/** 单条审批决策记录（service 层已脱敏） */
export interface TrackingRecord {
  actor_email: string;
  actor_name: string | null;
  action: string;
  reason: string;
  ts: string;
  /** 仅 admin 可见；申请人视角为 null */
  ip: string | null;
  /** 仅 admin 可见；申请人视角为 null */
  ua: string | null;
}

/** 实例申请人摘要 */
export interface TrackingApplicant {
  id: string;
  name: string | null;
}

/** 当前节点等待的 actor */
export interface TrackingCurrentActor {
  email: string | null;
  name: string | null;
}

/** 当前正在等待的节点（节点可视化要求） */
export interface TrackingCurrentNode {
  id: string;
  title: string | null;
  status: string;
  node_type: string;
  actor: TrackingCurrentActor | null;
  deadline_at: string | null;
  started_at: string | null;
}

/** 申请人追踪页响应 */
export interface TrackingResponse {
  instance_id: string;
  status: InstanceStatus;
  workflow_id: string;
  applicant: TrackingApplicant;
  current_node: TrackingCurrentNode | null;
  records: TrackingRecord[];
  created_at: string;
  completed_at: string | null;
}

/** HITL 节点状态徽章配置 — 节点可视化（user feedback 2026-05-17） */
export const HITL_NODE_STATUS_CONFIG: Record<
  string,
  { label: string; icon: string; className: string }
> = {
  pending: { label: '等待中', icon: '○', className: 'bg-gray-100 text-gray-600' },
  waiting_human: { label: '等待提交', icon: '⏳', className: 'bg-amber-100 text-amber-700' },
  in_review: { label: '审核中', icon: '👀', className: 'bg-blue-100 text-blue-700' },
  done: { label: '已通过', icon: '✓', className: 'bg-green-100 text-green-700' },
  rejected: { label: '已拒绝', icon: '✗', className: 'bg-red-100 text-red-700' },
  returned: { label: '已退回', icon: '↺', className: 'bg-orange-100 text-orange-700' },
  running: { label: '运行中', icon: '⟳', className: 'bg-blue-100 text-blue-700' },
  completed: { label: '已完成', icon: '✓', className: 'bg-green-100 text-green-700' },
  failed: { label: '失败', icon: '✗', className: 'bg-red-100 text-red-700' },
};

/** 决策 action 标签 */
export const RECORD_ACTION_LABELS: Record<string, { label: string; className: string }> = {
  submit: { label: '提交', className: 'bg-blue-50 text-blue-700' },
  approve: { label: '通过', className: 'bg-green-50 text-green-700' },
  reject: { label: '拒绝', className: 'bg-red-50 text-red-700' },
  return: { label: '退回', className: 'bg-orange-50 text-orange-700' },
  escalate: { label: '升级', className: 'bg-purple-50 text-purple-700' },
};
